### DISCUSSION TITLE
Non-stoichiometric reaction order issues

### DISCUSSION THREAD

**Discussion by JayStandridge** [AMReX-Combustion/PeleLMeX#287] [Q&A]
Hi all,

I have been studying various mechanisms, and recently have been taking a look at the Westbrook and Dryer two-step mechanism, as well as the Jones and Lindstedt four-step mechanism. Both mechanisms are currently supported by PelePhysics and come with the base installation of the Pele-suite codes. However, when using these mechanisms on the backward facing step base case of PeleLMeX, the chemical integrator gives bad results almost immediately. I have tried both ReactorCvode and ReactorRK64 to no avail, compiling with GNU. Does anyone have any experience running these mechanisms? 

Thank you in advance for your help.

**Comment by marchdf** [comment]
This might be a better discussion for PelePhysics but in any event @nickwimer and @malihass might have some thoughts.

**Comment by baperry2** [comment]
Tried running this in debug and with `amrex.fpe_trap_invalid = 1`, it raises an erroneous operation error on this line: https://github.com/AMReX-Combustion/PelePhysics/blob/71a77801e57fd9464ae91aadcaa7f54fd99e85e5/Support/Mechanism/Models/chem-CH4-2step/mechanism.H#L1876

Probably a species is getting a small negative value due diffusion/advection, and then the fractional exponent breaks. For FORD reactions we'll need to add clipping somewhere - perhaps directly where the fractional exponents are computed in the CEPTR-generated code.

**Reply by JayStandridge** [reply]
I tried running with peleLM.floor_species = 1 to see if species negativity is the issue, but that doesn't seem to fix the bug.

**Reply by baperry2** [reply]
It's possible the negative values enter the state (or the forcing for the reaction evaluation step) between when floorSpecies is called and when reactions are evaluated. As a test, try clipping the arguments to `pow` in the line of code linked above. Use  ReactorRK64 because the same issue would also crop up in the chemical Jacobian evaluation.

**Reply by marchdf** [reply]
I opened a PR in pelephysics to see if we can fix this https://github.com/AMReX-Combustion/PelePhysics/pull/433

**Reply by malihass** [reply]
I agree with Bruce, floor_species may not be sufficient here
could you please try this mechanism.H (https://github.com/AMReX-Combustion/PelePhysics/blob/clip_non_zero_pow/Support/Mechanism/Models/chem-CH4-2step/mechanism.H) ?
I also do the clipping for the Jacobian, so you should be able to use `AJdirect`

**Reply by JayStandridge** [reply]
This was the result when running with that mechanism.H:


====================   NEW TIME STEP   ==================== 
 Est. time step - Conv: 2.649152166e-14, divu: 7.437227909e-08
 STEP [0] - Time: 0, dt 2.649152166e-21
   SDC iter [1] 
amrex::Abort::amrex::Abort::1::MLMG failed !!!
amrex::Abort::2::MLMG failed !!!
amrex::Abort::3::MLMG failed !!!
amrex::Abort::4::MLMG failed !!!
amrex::Abort::5::MLMG failed !!!
amrex::Abort::amrex::Abort::7::MLMG failed !!!
0::MLMG failed !!!
SIGABRT

**Reply by marchdf** [reply]
hmmm can you try again? I made some (other) changes on my PR branch

**Reply by marchdf** [reply]
I merged that into main pelephysics. It might not be the end of the problems, but it was definitely an issue.
